sphinx/config.py

```diff
@@ -385,6 +385,15 @@ def __repr__(self) -> str:
             values.append(f"{opt_name}={opt_value!r}")
         return self.__class__.__qualname__ + '(' + ', '.join(values) + ')'
 
+    def __setattr__(self, key: str, value: Any) -> None:
+        # if someone is still using 'master_doc', we need to update 'root_doc'
```

Comment:
should we be trying to deprecate `master_doc`, i.e. should this emit a warning?

Reply:
also, just to note, there may be an edge case here where if both `master_doc` and `root_doc` are set one would silently override the other

Reply:
I would assume that if you put `master_doc` and `root_doc` in the same `conf.py`, then the latest change would be the one winning. For instance,

```python
master_doc = "1"
root_doc = "2"  # master_doc becomes 2
```

```python
root_doc = "2"
master_doc = "1"  # root_doc becomes 1
```

IMO, master_doc and root_doc are really the same, just aliased, so at any time, they should always have the same value (I should not expect master_doc != root_doc and should not be allowed to do it (well you could still bypass if you bypass `Config.__setattr__? , but at this point, you really want to mess with the API imo)).

Reply:
thats fine but do we still not want to deprecate `master_doc`, seems unnecessary to maintain both forever

Reply:
It's already deprecated since 4.0 but we are still allowing it to compile older projects actually I think. Otherwise, those old projects would require to be updated (I think, for some old projects, they are still compatible with the latest Sphinx *except* for maybe that configuration variable...).

Reply:
> It's already deprecated since 4.0

but without a deprecation warning, I feel people will never know to remove it,
so then we basically can never remove it no?

Reply:
I'll approve, to not hold it up, but yeh I definitely feel we should be nudging people in the right direction here

Reply:
>  so then we basically can never remove it no?

Technically,... it's possible but maybe hard to implement. The issue is that the deprecation is on an *attribute* and since we need to take into account that our modifications to that attribute are not deprecated but only those that are external, it's quite hard to detect them. Note that we can also do something like `config['...']` so it's even worse because it exposes another access...

Maybe using a dedicated setter but I'm not very happy to have a single property for that... (and again, detecting whether we are using it for backwards compatibility or not might be tricky, ... or we can just add a private method that does not trigger the warning?).